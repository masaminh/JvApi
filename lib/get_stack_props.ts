import { Construct } from 'constructs';
import JvApiStackProps from './jv_api_stack_props';

export default function getStackProps(app: Construct): JvApiStackProps {
  const jvdataBucket = app.node.tryGetContext('jvdataBucket');
  const jvdataPrefix = app.node.tryGetContext('jvdataPrefix');
  const stage = app.node.tryGetContext('stage');
  const stackName = app.node.tryGetContext('stackName');

  if (typeof jvdataBucket !== 'string' || typeof jvdataPrefix !== 'string' || typeof stage !== 'string') {
    throw new Error('jvdataBucket and jvdataPrefix and stage must be set in context');
  }

  if (stackName !== undefined && typeof stackName !== 'string') {
    throw new Error('stackName must be string');
  }

  return {
    jvdataBucket, jvdataPrefix, stage, stackName,
  };
}
