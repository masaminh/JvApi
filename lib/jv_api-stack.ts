import { Stack, Duration, RemovalPolicy } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { Architecture, Runtime } from 'aws-cdk-lib/aws-lambda';
import { NodejsFunction } from 'aws-cdk-lib/aws-lambda-nodejs';
import { LogGroup, RetentionDays } from 'aws-cdk-lib/aws-logs';
import { Bucket } from 'aws-cdk-lib/aws-s3';
import { StringParameter } from 'aws-cdk-lib/aws-ssm';
import JvApiStackProps from './jv_api_stack_props';

// eslint-disable-next-line import/prefer-default-export
export class JvApiStack extends Stack {
  constructor(scope: Construct, id: string, props: JvApiStackProps) {
    super(scope, id, props);

    const logGroup = new LogGroup(this, 'ApiFunctionLogGroup', {
      logGroupName: `${this.stackName}/ApiFunction`,
      retention: RetentionDays.ONE_MONTH,
      removalPolicy: RemovalPolicy.DESTROY,
    });

    const apiFunction = new NodejsFunction(this, 'ApiFunction', {
      entry: 'src/handler.ts',
      handler: 'handler',
      runtime: Runtime.NODEJS_18_X,
      architecture: Architecture.ARM_64,
      timeout: Duration.seconds(30),
      environment: {
        JVDATA_BUCKET: props.jvdataBucket,
        JVDATA_PREFIX: props.jvdataPrefix,
      },
      logGroup,
    });

    const url = apiFunction.addFunctionUrl();

    const s3 = Bucket.fromBucketName(this, 'CacheBucket', props.jvdataBucket);

    s3.grantRead(apiFunction);

    // eslint-disable-next-line no-new
    new StringParameter(this, 'ApiUrlParameter', {
      parameterName: `/${props.stage}/JvApi/Url`,
      stringValue: url.url,
    });

    // eslint-disable-next-line no-new
    new StringParameter(this, 'FunctionArnParameter', {
      parameterName: `/${props.stage}/JvApi/FunctionArn`,
      stringValue: apiFunction.functionArn,
    });
  }
}
