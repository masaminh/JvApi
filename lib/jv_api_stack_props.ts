import { StackProps } from 'aws-cdk-lib';

interface JvApiStackProps extends StackProps {
  jvdataBucket: string;
  jvdataPrefix: string;
  stage: string;
}

export default JvApiStackProps;
