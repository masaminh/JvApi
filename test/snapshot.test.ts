import * as cdk from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { JvApiStack } from '../lib/jv_api-stack';

test('snapshot test', () => {
  const app = new cdk.App();

  const stack = new JvApiStack(app, 'MyTestStack', {
    env: { region: 'ap-northeast-1' },
    jvdataBucket: 'jvbucket',
    jvdataPrefix: 'JVDATAPREFIX',
    stage: 'STAGE',
  });

  // スタックからテンプレート(JSON)を生成
  const template = Template.fromStack(stack).toJSON();

  // 生成したテンプレートとスナップショットが同じか検証
  expect(template).toMatchSnapshot();
});
