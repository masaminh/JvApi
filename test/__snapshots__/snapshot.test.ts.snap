// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "ApiFunctionCE271BD4": {
      "DependsOn": [
        "ApiFunctionServiceRoleDefaultPolicy20A32B8D",
        "ApiFunctionServiceRole52B9747B",
      ],
      "Properties": {
        "Architectures": [
          "arm64",
        ],
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-ap-northeast-1",
          },
          "S3Key": "HASH-REPLACED.zip",
        },
        "Environment": {
          "Variables": {
            "JVDATA_BUCKET": "jvbucket",
            "JVDATA_PREFIX": "JVDATAPREFIX",
          },
        },
        "Handler": "index.handler",
        "LoggingConfig": {
          "LogGroup": {
            "Ref": "ApiFunctionLogGroup00D1AC73",
          },
        },
        "Role": {
          "Fn::GetAtt": [
            "ApiFunctionServiceRole52B9747B",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
        "Timeout": 30,
        "TracingConfig": {
          "Mode": "Active",
        },
      },
      "Type": "AWS::Lambda::Function",
    },
    "ApiFunctionFunctionUrl73AD62DC": {
      "Properties": {
        "AuthType": "NONE",
        "TargetFunctionArn": {
          "Fn::GetAtt": [
            "ApiFunctionCE271BD4",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Url",
    },
    "ApiFunctionLogGroup00D1AC73": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "LogGroupName": "MyTestStack/ApiFunction",
        "RetentionInDays": 30,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Delete",
    },
    "ApiFunctionServiceRole52B9747B": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "ApiFunctionServiceRoleDefaultPolicy20A32B8D": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "xray:PutTraceSegments",
                "xray:PutTelemetryRecords",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::jvbucket",
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::jvbucket/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "ApiFunctionServiceRoleDefaultPolicy20A32B8D",
        "Roles": [
          {
            "Ref": "ApiFunctionServiceRole52B9747B",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "ApiFunctioninvokefunctionurl0B8A0798": {
      "Properties": {
        "Action": "lambda:InvokeFunctionUrl",
        "FunctionName": {
          "Fn::GetAtt": [
            "ApiFunctionCE271BD4",
            "Arn",
          ],
        },
        "FunctionUrlAuthType": "NONE",
        "Principal": "*",
      },
      "Type": "AWS::Lambda::Permission",
    },
    "ApiUrlParameterF80EB5D4": {
      "Properties": {
        "Name": "/STAGE/JvApi/Url",
        "Type": "String",
        "Value": {
          "Fn::GetAtt": [
            "ApiFunctionFunctionUrl73AD62DC",
            "FunctionUrl",
          ],
        },
      },
      "Type": "AWS::SSM::Parameter",
    },
    "FunctionArnParameter94E55A91": {
      "Properties": {
        "Name": "/STAGE/JvApi/FunctionArn",
        "Type": "String",
        "Value": {
          "Fn::GetAtt": [
            "ApiFunctionCE271BD4",
            "Arn",
          ],
        },
      },
      "Type": "AWS::SSM::Parameter",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
